### PR TITLE
Add torch support to ParametricUMAP

### PR DESCRIPTION
**Pull request by fchollet**
This is a follow-up CL to #1101. It makes the code backend-agnostic and unlocks PyTorch support. It also removes some amount of unused code as well as a couple of constructor arguments identified as unnecessary.

Notes:

- This restricts support to Keras 3 only. IMO this is not a problem because Keras 2 is now considered legacy, and those who need to use it can simply stick to an earlier version of umap-learn. Supporting both backend-agnostic Keras 3 AND Keras 2 would be complex. It's easy to support either {Keras 2, Keras 3 + TF} or {Keras 3 + any backend}, but not the union of both sets.
- TF is still needed for the tf.data pipeline. We could change that in the future, though. There's no actual need to use tf.data.
- JAX isn't yet supported, due to one line in the umap loss that calls a RNG function (shuffle). JAX is stateless so the RNG call needs to be made stateless. We can do this easily via the model-centric refactor proposed in https://github.com/lmcinnes/umap/pull/1101#issuecomment-2005717945.
- In order not to break PyTorch support going forward, you would need to start running the ParametricUMAP tests with the PyTorch backend too. Right now the PyTorch backend won't be tested on CI so it might break upon future changes.
- We still need to perform the refactors from https://github.com/lmcinnes/umap/pull/1101#issuecomment-2005717945 for code health.

### CODE REVIEW

**Comment by lmcinnes** [comment]
It looks like this is just issues with py3.8 which I presume Keras 3 does not necessarily support. I'm happy to move on on that front. It looks good to me, but I would appreciate it if @timsainb could give it a once over before I hit merge.

**Comment by timsainb** [comment]
Thanks all! I’m on a trip now and will take a look as soon as I get a chance
